File: Algorithm I/3. Sliding Window - Day 6/P3.cpp
// Date: Tue, Sep 28, 2021
// 3. Longest Substring Without Repeating Characters (https://leetcode.com/problems/permutation-in-string/)

// sliding window technique
// time complexity O(n)
// space complexity O(d)
int lengthOfLongestSubstring(string s) {
	int n = s.size();
	int ans = 0, begin = 0;
	vector<int> last_occurrence(256, -1);
	for (int end = 0; end < n; end++) {
		int last_seen_index = last_occurrence[s[end]];
		if (last_seen_index >= begin)
			begin = last_seen_index + 1;
		last_occurrence[s[end]] = end;
		ans = max(ans, end - begin + 1);
	}
	return ans;
}

File: LeetCoding-Challenge/June-2022/P3.java
package leetcodingchallenge.june2022;

import java.util.HashMap;
import java.util.Map;

/**
 * Date: June 10, 2022 (just redo, already solve it before in C++)
 * 3. Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 */

public class P3 {
    // time: O(n)
    // space: O(n)
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> hash = new HashMap<>();
        int n = s.length(), ans = 0, first = 0;
        for (int i = 0; i < n; i++) {
            char ch = s.charAt(i);
            if (hash.containsKey(ch) && hash.get(ch) >= first) {
                first = hash.get(ch) + 1;
            } else {
                ans = Math.max(ans, i - first + 1);
            }
            hash.put(s.charAt(i), i);
        }
        return ans;
    }
}
